### PR TITLE
Use `BTreeSet` instead of `Vec`/`HashSet` for storing validators who have seen an Ethereum event

### PR DESCRIPTION
**Pull request by james-chf**
Closes https://github.com/anoma/namada/issues/434

Updates the spec and code to:
- update specs to specify a `BTreeSet<Address>` instead of `Vec<Address>` for the `eth_msgs/$msg_hash/seen_by` field
- update `EthMsg::seen_by` to be a `BTreeSet`
- update `MultiSignedEthEvent::signers` to be a `BTreeSet`

This is a better data structure for doing https://github.com/anoma/namada/issues/197 (updating Ethereum events in storage) as when new validators see an event, we can read the `seen_by` from storage and insert new validators without having to worry about reordering a `Vec` before we store it again. We also get a uniqueness constraint on validators (the elements of the `BTreeSet`) which `Vec` doesn't give us.

### CODE REVIEW

File: shared/src/types/vote_extensions/ethereum_events.rs [hunk 1]
```diff
@@ -79,11 +79,11 @@ pub struct MultiSignedEthEvent {
...
     /// List of addresses of validators who signed this event
     #[cfg(feature = "abcipp")]
-    pub signers: HashSet<Address>,
+    pub signers: BTreeSet<Address>,
     /// List of addresses of validators who signed this event
     /// and block height at which they signed it
     #[cfg(not(feature = "abcipp"))]
-    pub signers: HashSet<(Address, BlockHeight)>,
+    pub signers: BTreeSet<(Address, BlockHeight)>,
 }
```
Comment (james-chf) [inline shared/src/types/vote_extensions/ethereum_events.rs:87]: `MultiSignedEthEvent` and `EthMsgUpdate` are now the same for ABCI++ and may become the same for ABCI+ as well soon. These structs could probably be combined in the future - a `MultiSignedEthEvent` translates straightfowardly into an `EthMsgUpdate`.

**Comment by james-chf** [comment]
pls update wasm

**Review by sug0 — APPROVED** [review]

**Comment by james-chf** [comment]
pls update wasm

**Comment by james-chf** [comment]
pls update wasm

**Pull request merged**